test/UnitTests/MSTest.Analyzers.UnitTests/DataRowShouldBeValidAnalyzerTests.cs

```diff
@@ -0,0 +1,325 @@
+    {
+        var code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            [TestClass]
+            public class MyTestClass
+            {
+                [DataRow(1)]
+                public void {|#0:TestMethod1|}(int i)
+                {
+                }
+            }
+            """
+        ;
+
+        await VerifyCS.VerifyAnalyzerAsync(code, VerifyCS.Diagnostic(DataRowShouldBeValidAnalyzer.DataRowOnTestMethodRule).WithLocation(0));
+    }
+
+    public async Task WhenDataRowHasNoArgs_Diagnostic()
+    {
+        var code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            [TestClass]
+            public class MyTestClass
+            {
+                [{|#0:DataRow()|}]
+                [TestMethod]
+                public void TestMethod1(int i)
+                {
+                }
+            }
+            """
+        ;
+
+        await VerifyCS.VerifyAnalyzerAsync(code, VerifyCS.Diagnostic(DataRowShouldBeValidAnalyzer.AtLeastOneArgumentRule).WithLocation(0));
+    }
+
+    public async Task WhenDataRowHasArgumentMismatchWithTestMethod_Diagnostic()
+    {
+        var code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            [TestClass]
+            public class MyTestClass
+            {
+                [{|#0:DataRow(1, 2, 3)|}]
+                [TestMethod]
+                public void TestMethod1(int i, int j, int k, int l)
+                {
+                }
+            }
+            """
+        ;
+
+        await VerifyCS.VerifyAnalyzerAsync(code, VerifyCS.Diagnostic(DataRowShouldBeValidAnalyzer.ArgumentNumberRule).WithLocation(0));
+    }
+
+    public async Task WhenDataRowHasArgumentMismatchWithTestMethod2_Diagnostic()
+    {
+        var code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            [TestClass]
+            public class MyTestClass
+            {
+                [{|#0:DataRow(1, 2, 3)|}]
+                [TestMethod]
+                public void TestMethod1(int i, int j)
+                {
+                }
+            }
+            """
+        ;
+
+        await VerifyCS.VerifyAnalyzerAsync(code, VerifyCS.Diagnostic(DataRowShouldBeValidAnalyzer.ArgumentNumberRule).WithLocation(0));
+    }
+
+    public async Task WhenDataRowHasTypeMismatchWithTestMethod_Diagnostic()
+    {
+        var code = """
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+            [TestClass]
+            public class MyTestClass
+            {
+                [{|#0:DataRow(1, 2, 3)|}]
+                [TestMethod]
+                public void TestMethod1(int i, int j, string s)
+                {
+                }
+            }
+            """
+        ;
+
+        await VerifyCS.VerifyAnalyzerAsync(code, VerifyCS.Diagnostic(DataRowShouldBeValidAnalyzer.ArgumentTypeRule).WithLocation(0));
+    }
+}
```

Comment:
It'd be great to have one test with `[DataTestMethod]` instead of `[TestMethod]` and another with a custom local attribute that inherits from `TestMethodAttribute`.

Reply:
Done, thanks for the suggestions !